In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
#Importing the data and stored inside path variables
path = '/content/drive/MyDrive/samplecar (1).csv'

In [ ]:
df = pd.read_csv(path)
print(df.shape)

(10, 3)


In [ ]:
df.head()

Type Price  Buy
0    SUV     H   No
1     HB     H   No
2     HB     H  Yes
3     HB     L   No
4  SEDAN     H  Yes

In [ ]:
df.describe()

Type Price  Buy
count    10    10   10
unique    3     3    2
top      HB     H  Yes
freq      4     5    6

In [ ]:
#finding the unique value of each columns
for col in df.columns:
    print("unique value of col {} are {}".format(col,df[col].unique()))

unique value of col Type are ['SUV' 'HB' 'SEDAN']
unique value of col Price are ['H' 'L' 'Hot']
unique value of col Buy are ['No' 'Yes']


Now once we have find the unique value in each column thenafter we will calcuate the total entropy for each initial state.

In [ ]:
def calEntropy(totalLen,totalLenA,totalLenB):
    pobA = totalLenA/totalLen
    pobB = totalLenB/totalLen
    ent = -pobA * np.log2(pobA) - pobB*np.log2(pobB)
    return ent

The above fuction is used to calculate the entropy for all the attributes and combinations. 

#Now calculating the entropy of type buy using the above function. 

In [ ]:
tot = len(df)
tot_n = len(df[df['Buy']=="No"])
tot_y = len(df[df['Buy']=="Yes"])
prob_y = tot_y/tot
prob_n = tot_n/tot
print(prob_y, prob_n)
H_tot = -prob_y * np.log2(prob_y) - prob_n * np.log2(prob_n)
print(H_tot)

0.6 0.4
0.9709505944546686


#Calculating the entropy of each type features

# At first we calculate the entropy of type SUV

In [ ]:
suv = df[df['Type']=='SUV']
suv_len = len(suv)
print("Total len of SUV= ",suv_len)
suv_and_yes = len(suv[suv['Buy']=='Yes'])
print("Total len of SUV and Yes= ",suv_and_yes)
suv_and_no = len(suv[suv['Buy']=='No'])
print("Total len of SUV and No=",suv_and_no)
ent_suv = calEntropy(suv_len,suv_and_yes,suv_and_no)## Here we have call the above entropy fuction to calculate the entrpy of SUV type
print("Entropy of SUV = ",ent_suv)

Total len of SUV=  3
Total len of SUV and Yes=  1
Total len of SUV and No= 2
Entropy of SUV =  0.9182958340544896


# Similarly, we calculate the entropy of type HB

In [ ]:
hb = df[df['Type']=='HB']
hb_len = len(hb)
print("Total len of HB= ",hb_len)
hb_and_yes = len(hb[hb['Buy']=='Yes'])
print("Total len of HB and Yes",hb_and_yes)
hb_and_no = len(hb[hb['Buy']=='No'])
print("Total len of HB and No",hb_and_no)
ent_hb = calEntropy(hb_len,hb_and_yes,hb_and_no)
print("Entropy of HB = ",ent_hb)

Total len of HB=  4
Total len of HB and Yes 2
Total len of HB and No 2
Entropy of HB =  1.0


# Lastly, aagain we calculate entropy of type SEDAN

In [ ]:
sedan = df[df['Type']=='SEDAN']
sedan_len = len(sedan)
print("Total len of SEDAN= ",sedan_len)
sedan_and_yes = len(sedan[sedan['Buy']=='Yes'])
print("Total len of SEDAN and Yes= ",sedan_and_yes)
sedan_and_no = len(sedan[sedan['Buy']=='No'])
print("Total len of SEDAN and No= ",sedan_and_no)
ent_sedan = calEntropy(sedan_len,sedan_and_yes,sedan_and_no)
print("Entropy of SEDAN = ",ent_sedan)

Total len of SEDAN=  3
Total len of SEDAN and Yes=  3
Total len of SEDAN and No=  0
Entropy of SEDAN =  nan


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: RuntimeWarning: divide by zero encountered in log2
  after removing the cwd from sys.path.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: RuntimeWarning: invalid value encountered in double_scalars
  after removing the cwd from sys.path.


Once we have find the entropy of each unique value, then we will calculate the overall entropy of the type. 

In [ ]:
Overall_Entropy_Of_Type = (3/10)*0.9182958340544896 + (4/10)*1 + (3/10)*0
Overall_Entropy_Of_Type

0.6754887502163469

# Now we have overall entropy of type attribute, then after we will calculate the Information Gain for the type attribute. 
#For this we have the formula
   \begin{align}
        \mathbf{Entropy(T,X)} =  [\sum_{c=X} p(c)E(c) ]
    \end{align}

#OR For this task we can say

Information Gain = Entrpy of type buy - Entropy of type 

In [ ]:
InformationGain_Type = 0.9709505944546686 - 0.6754887502163469
InformationGain_Type

0.29546184423832167

# Similarly, again we follow the same step for calculating the entropy of each price features

At first we calculate the entropy of price H

In [ ]:
h = df[df['Price']=='H']
h_len = len(h)
print("Total len of H= ",h_len)
h_and_yes = len(h[h['Buy']=='Yes'])
print("Total len of H and Yes= ",h_and_yes)
h_and_no = len(h[h['Buy']=='No'])
print("Total len of H and No= ",h_and_no)
ent_h = calEntropy(h_len,h_and_yes,h_and_no)
print("Entropy of H = ",ent_h)

Total len of H=  5
Total len of H and Yes=  3
Total len of H and No=  2
Entropy of H =  0.9709505944546686


Calculating the entropy of price L

In [ ]:
l = df[df['Price']=='L']
l_len = len(l)
print("Total len of L= ", l_len)
l_and_yes = len(l[l['Buy']=='Yes'])
print("Total len of L and Yes= ",l_and_yes)
l_and_no = len(l[l['Buy']=='No'])
print("Total len of L and No= ",l_and_no)
ent_l = calEntropy(l_len,l_and_yes,l_and_no)
print("Entropy of L = ",ent_l)

Total len of L=  4
Total len of L and Yes=  2
Total len of L and No=  2
Entropy of L =  1.0


Calculating the entropy of price Hot

In [ ]:
hot = df[df['Price']=='Hot']
hot_len = len(hot)
print("Total len of Hot= ", hot_len)
hot_and_yes = len(hot[hot['Buy']=='Yes'])
print("Total len of Hot and Yes= ",hot_and_yes)
hot_and_no = len(hot[hot['Buy']=='No'])
print("Total len of Hot and No= ",hot_and_no)
ent_hot = calEntropy(hot_len,hot_and_yes,hot_and_no)
print("Entropy of Hot = ",ent_hot)

Total len of Hot=  1
Total len of Hot and Yes=  1
Total len of Hot and No=  0
Entropy of Hot =  nan


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: RuntimeWarning: divide by zero encountered in log2
  after removing the cwd from sys.path.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: RuntimeWarning: invalid value encountered in double_scalars
  after removing the cwd from sys.path.


In [ ]:
Overall_Entropy_Of_Price= (5/10)*0.9709505944546686 + (4/10)*1 + (1/10)*0
Overall_Entropy_Of_Price

0.8854752972273343

In [ ]:
InformationGain_Price = 0.9709505944546686 - 0.8854752972273343
InformationGain_Price

0.08547529722733427

# Conclusion

From the above calculation we saw that information gain of type is higher that information gain of price.

InformationGain_Type > InformationGain_Price

i.e. 0.29546184423832167 > 0.08547529722733427

